src/v/datalake/translation/deps.cc

```diff
@@ -10,12 +10,519 @@
 
 #include "datalake/translation/deps.h"
 
+#include "cluster/notification.h"
+#include "cluster/partition.h"
+#include "datalake/coordinator/frontend.h"
+#include "datalake/local_parquet_file_writer.h"
+#include "datalake/record_schema_resolver.h"
+#include "datalake/record_translator.h"
+#include "datalake/serde_parquet_writer.h"
+#include "datalake/translation/state_machine.h"
+#include "datalake/translation_task.h"
+#include "kafka/data/partition_proxy.h"
+#include "kafka/utils/txn_reader.h"
+
 namespace datalake::translation {
 
+namespace {
+ss::future<cluster::errc> wait_stm_translated(
+  ss::shared_ptr<translation_stm> stm,
+  model::offset o,
+  model::timeout_clock::time_point deadline,
+  std::optional<std::reference_wrapper<ss::abort_source>> as) {
+    try {
+        co_await stm->wait_translated(model::prev_offset(o), deadline, as);
+    } catch (const ss::abort_requested_exception&) {
+        co_return cluster::errc::shutting_down;
+    } catch (const ss::timed_out_error&) {
+        co_return cluster::errc::timeout;
+    }
+    co_return cluster::errc::success;
+}
+} // namespace
+
+static constexpr auto poll_duration = 2s;
+
 ss::future<> noop_mem_tracker::maybe_reserve_memory(size_t, ss::abort_source&) {
     return ss::make_ready_future<>();
 }
 void noop_mem_tracker::update_current_memory_usage(size_t) {}
 void noop_mem_tracker::release() {}
 
+ss::future<> writer_reservations_impl::maybe_reserve_memory(
+  size_t bytes, ss::abort_source& as) {
+    while (_available_memory < bytes) {
```

Comment:
is it possible that that `_available_memory` will always be smaller than `byte` ?

Reply:
you mean in a situation where `bytes` exceeds the total memory in `reservations_tracker`? I think in that case 
- the scheduler gets notified of resource exhaustion
- the reservations_tracker::reserve_memory call is put into abortable sleep
- the scheduler stops all translators, each of which triggers its inflight abort source.
- which causes reserve_memory to throw sleep aborted
- which breaks the loop and lands in parquet_writer, which fails gracefully

so I don't _think_ so? do you see another outcome?

Reply:
Can this lead to a case where nothing makes progress? I.e. N translations are scheduled, all of them try to acquire memory and do some writing but none of them ever acquire enough memory to complete a write? I see that on `maybe_reserve_memory` we just fail the write and restart.

Reply:
In that case scheduler kicks in and aborts the translator using most memory so others can make progress?

Reply:
Ah. @oleiman said it a bit differently

> the scheduler stops **all** translators, each of which triggers its inflight abort source.

The way you put it makes a bit more sense.

Reply:
hmm, yeah good shout. not sure where I got that from. overall point being that `reservations_tracker::reserve_memory` will notify the scheduler if it can't service the reservation - maybe that should be captured in a comment somewhere?